api4/channel.go

```diff
@@ -1753,3 +1761,214 @@ func patchChannelModerations(c *Context, w http.ResponseWriter, r *http.Request)
+		c.SetPermissionError(model.PERMISSION_EDIT_OTHER_USERS)
+		return
+	}
+
+	categories, err := c.App.GetSidebarCategory(c.Params.UserId, c.Params.TeamId, c.Params.CategoryId)
+	if err != nil {
+		c.Err = err
+		return
+	}
+
+	w.Write(categories.ToJson())
+}
+
+func updateCategoriesForTeamForUser(c *Context, w http.ResponseWriter, r *http.Request) {
+	c.RequireUserId().RequireTeamId().RequireCategoryId()
+	if c.Err != nil {
+		return
+	}
+
+	if !c.App.SessionHasPermissionToUser(*c.App.Session(), c.Params.UserId) || !c.App.SessionHasPermissionToCategory(*c.App.Session(), c.Params.TeamId, c.Params.CategoryId) {
+		c.SetPermissionError(model.PERMISSION_EDIT_OTHER_USERS)
+		return
+	}
+
+	auditRec := c.MakeAuditRecord("updateCategoriesForTeamForUser", audit.Fail)
+	defer c.LogAuditRec(auditRec)
+
+	categoriesUpdateRequest, err := model.SidebarCategoriesFromJson(r.Body)
+	if err != nil {
+		c.SetInvalidParam("category")
+		return
+	}
+	var channelsToCheck []string
+	for _, category := range categoriesUpdateRequest {
+		channelsToCheck = append(channelsToCheck, category.Channels...)
+	}
+	if appErr := validateUserChannels("updateCategoriesForTeamForUser", c, c.Params.TeamId, c.Params.UserId, channelsToCheck); appErr != nil {
+		c.Err = appErr
+		return
+	}
+
+	categories, appErr := c.App.UpdateSidebarCategories(c.Params.UserId, c.Params.TeamId, categoriesUpdateRequest)
+	if appErr != nil {
+		c.Err = appErr
+		return
+	}
+
+	auditRec.Success()
+	w.Write(model.SidebarCategoryWithChannelsToJson(categories))
+}
+
+func validateUserChannels(operationName string, c *Context, teamId, userId string, channelIDs []string) *model.AppError {
+	channels, err := c.App.GetChannelsForUser(teamId, userId, false)
+	if err != nil {
+		return model.NewAppError("Api4."+operationName, "api.invalid_channel", nil, err.Error(), http.StatusInternalServerError)
+	}
+	for _, channelId := range channelIDs {
+		found := false
+		for _, channel := range *channels {
+			if channel.Id == channelId {
+				found = true
+				break
+			}
+			if !found {
+				return model.NewAppError("Api4."+operationName, "api.invalid_channel", nil, err.Error(), http.StatusInternalServerError)
+			}
+		}
+	}
+	return nil
+}
+
+func updateCategoryForTeamForUser(c *Context, w http.ResponseWriter, r *http.Request) {
+	c.RequireUserId().RequireTeamId().RequireCategoryId()
+	if c.Err != nil {
+		return
+	}
+
+	if !c.App.SessionHasPermissionToUser(*c.App.Session(), c.Params.UserId) || !c.App.SessionHasPermissionToCategory(*c.App.Session(), c.Params.TeamId, c.Params.CategoryId) {
+		c.SetPermissionError(model.PERMISSION_EDIT_OTHER_USERS)
+		return
+	}
+
+	auditRec := c.MakeAuditRecord("updateCategoryForTeamForUser", audit.Fail)
+	defer c.LogAuditRec(auditRec)
+
+	categoryUpdateRequest, err := model.SidebarCategoryFromJson(r.Body)
+	if err != nil {
+		c.SetInvalidParam("category")
+		return
+	}
```

Comment:
As above, need to check userId, teamId and categoryId are equal in body payload and in URL parameters.